sources/api/migration/migrations/v1.0.5/add-user-data/src/main.rs

```diff
@@ -0,0 +1,45 @@
+#![deny(rust_2018_idioms)]
+
+use migration_helpers::{migrate, Migration, MigrationData, Result};
+use std::process;
+
+/// This migration removes host-container user data settings when downgrading to versions that
+/// don't understand them.
+pub struct AddUserDataMigration;
+
+impl Migration for AddUserDataMigration {
+    /// There's no user data by default, it's just left empty on upgrade.
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        println!("AddUserDataMigration has no work to do on upgrade.");
+        Ok(input)
+    }
+
+    /// Older versions don't know about the user-data settings; we remove them so that old versions
+    /// don't see them and fail deserialization.
+    fn backward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
```

Comment:
Is there any reason to delete the user-data file(s) that might exist?

Reply:
I don't _think_ so, but it's an interesting question to leave up for others to consider.

It seems to me that the worst case might be downgrading and still using a host container that understands the user-data file but tries to do something with it that isn't compatible with the older Bottlerocket version..?  For example, maybe you're using the user-data to configure a host-container that makes API calls back to the parent, and those calls are no longer valid with the downgraded version.  I think, in that case, you would have created an implicit dependency from the container to the host version, and so should remove/downgrade the host container before downgrading Bottlerocket.

I'm hesitant to delete things from host container persistent storage, so I'd like to have a decent reason, or fall back to leaving what (naively) seems to be a harmless file.

Reply:
It's a really interesting question and one that has some pretty good arguments both ways.  For now I think we should leave the file there.

Reply:
I agree, I think it's far enough out there as an edge case that it doesn't justify deleting data from the persistent store.